### ISSUE TITLE
Cancel order and restore quote methods increase stocks twice

### ISSUE THREAD

**Issue by simpleadm** [magento/magento2#9969]
Some payment methods in cancel payment action try to cancel current order and restore quote and as the result product qty is increased by 2.

### Affected payment methods
1. Paypal Hosted Pro (CE, EE)
2. Paypal Payflow (CE, EE)
3. Worldpay (EE)
4. Eway (EE)
5. PayEx (Third party payment gateway)

### Preconditions
1. Magento 2.1.* CE and EE
2. Configured payment methods

### Steps to reproduce
1. Suppose you have product with stock quantity: 1
2. Add product to cart and Checkout
3. Select affected payment method
4. The order is placed and stock is reduced to 0
5. Click cancel on Payment gateway (e.g. press "Cancel and return to store" button on PayPal page) 

### Expected result
1. Order was canceled
2. Product stock quantity: 1

### Actual result
1. Order was canceled
2. Product stock quantity: 2 (increase to 2)

### Additional information
For example Paypal Hosted Pro Cancel action 
[\Magento\Paypal\Controller\Hostedpro\Cancel](https://github.com/magento/magento2/blob/develop/app/code/Magento/Paypal/Controller/Hostedpro/Cancel.php)
```php
   /**
     * Customer canceled payment on gateway side.
     *
     * @return void
     */
    public function execute()
    {
        $this->checkoutHelper->cancelCurrentOrder('');
        $this->checkoutHelper->restoreQuote();

        $this->_redirect('checkout', ['_fragment' => 'payment']);
    }
```

Method `cancelCurrentOrder` will execute `$order->registerCancellation($comment)->save();` that will cancel all order items and dispatch 'sales_order_item_cancel' event for each one. `\Magento\CatalogInventory\Observer\CancelOrderItemObserver` observer listens this  event and increases product qty.

Next `\Magento\Checkout\Model\Session::restoreQuote()` will be executed to restore customer quote. That method will update and save a quote and dispatch `restore_quote` event.
`\Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver` listens to this event and product qty is increased the second time.

The problem is that both statements increase quantity of product.

The same issue in Payflow cancel action `\Magento\Paypal\Controller\Payflow`:
```php
     /**
     * Cancel order, return quote to customer
     *
     * @param string $errorMsg
     * @return false|string
     */
    protected function _cancelPayment($errorMsg = '')
    {
        $errorMsg = trim(strip_tags($errorMsg));

        $gotoSection = false;
        $this->_checkoutHelper->cancelCurrentOrder($errorMsg);
        if ($this->_checkoutSession->restoreQuote()) {
            //Redirect to payment step
            $gotoSection = 'paymentMethod';
        }

        return $gotoSection;
    }
```

**Comment by biclope** [comment]
you found it, you patched it

**Comment by IlnitskiyArtem** [comment]
@simpleadm, Thanks for reporting this issue. 
We've created internal ticket MAGETWO-70440 to address this issue.

**Comment by ttraskback** [comment]
Would it not work to just check in `\Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver::execute` if order is canceled don't add items back to stock?
```php
$order = $observer->getEvent()->getOrder();
if($order->getState() == \Magento\Sales\Model\Order::STATE_CANCELED){
    // Clear flag, so if order placement retried again with success - it will be processed
    $quote->setInventoryProcessed(false);
    return;
}
```

**Comment by simpleadm** [comment]
@ttraskback it should work.

Currently will be better to add following plugin:

_di.xml_
```xml
<type name="Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver">
        <plugin name="CatalogInventoryPlugin::RevertQuoteInventoryObserver" type="VendorName\ModuleName\Plugin\CatalogInventory\RevertQuoteInventoryObserver" sortOrder="100" />
</type>
```

_RevertQuoteInventoryObserver.php_
```php
    /**
     * Skip revert products sale for canceled orders
     *
     * @param \Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver $subject
     * @param \Closure $proceed
     * @param \Magento\Framework\Event\Observer $observer
     * @return void
     * @see \Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver::execute
     */
    public function aroundExecute(
        \Magento\CatalogInventory\Observer\RevertQuoteInventoryObserver $subject,
        \Closure $proceed,
        \Magento\Framework\Event\Observer $observer
    ) {
        /** @var \Magento\Sales\Model\Order $order */
        $order = $observer->getOrder();

        if ($order && $order->isCanceled()) {
            return;
        }

        $proceed($observer);
    }
```

**Comment by ttraskback** [comment]
Hi @simpleadm,
Yea i also made a plugin to fix the problem in our projects.
Tough i still do the `$quote->setInventoryProcessed(false);` so that the buy process executes correctly.

**Comment by magento-engcom-team** [comment]
@simpleadm, thank you for your report.
We've created internal ticket(s) MAGETWO-70440 to track progress on the issue.

**Comment by adrian-martinez-interactiv4** [comment]
For anyone interested: https://github.com/magento/magento2/pull/12668#issuecomment-354443814

**Comment by magento-team** [comment]
Hi @simpleadm. Thank you for your report. 
The issue has been fixed in magento/magento2#12668 by @dverkade in 2.2-develop branch
Related commit(s):
 - 2c1d6a4d115f1e97787349849d215e6c73ac1335

The fix will be available with the upcoming patch release.

**Comment by magento-team** [comment]
Hi @simpleadm. Thank you for your report. 
The issue has been fixed in magento/magento2#12949 by @dverkade in 2.3-develop branch
Related commit(s):
 - 10550760af088f5325202e19e8a8371c8c5072a2

The fix will be available with the upcoming patch release.

**Comment by magento-team** [comment]
Hi @simpleadm. Thank you for your report. 
The issue has been fixed in magento/magento2#12952 by @dverkade in 2.1-develop branch
Related commit(s):
 - bbe2f1a4b2ade0403760196c001d1b7156a5ddcb
 - b5b59be6fa3676a98d7777abfdc01c525ba587ce

The fix will be available with the upcoming patch release.
